File: client/lib/adminPlayers.ts
import { isAdminEmail } from "./admin";
import { getPlayers } from "./playerStore";
import { compareCompetitiveStanding, rankPlayers } from "./ranking";
import type { GameStatus, Player } from "./types";

type AdminRound = {
  round: number;
  attempts: number;
  score: number;
  prompt: unknown;
  output: string;
};

export type RawAdminPlayerDoc = {
  _id?: unknown;
  name?: unknown;
  email?: unknown;
  roundsPlayed?: unknown;
  timeTaken?: unknown;
  avgAccuracy?: unknown;
  attemptsTaken?: unknown;
  gameStatus?: unknown;
  createdAt?: unknown;
  completedAt?: unknown;
  rounds?: unknown;
};

export type AdminPlayerSummary = {
  playerId: string;
  name: string;
  email?: string;
  roundsPlayed: number;
  timeTakenSec: number;
  averageScore: number;
  attemptsUsed: number;
  completed: boolean;
  gameStatus?: string;
};

export type AdminPlayerExport = {
  _id: string;
  name: string;
  email?: string;
  roundsPlayed: number;
  timeTaken: number;
  avgAccuracy: number;
  attemptsTaken: number;
  gameStatus: string;
  createdAt?: string;
  completedAt?: string;
  rounds: AdminRound[];
};

type NormalizedAdminPlayer = {
  id: string;
  name: string;
  email?: string;
  roundsPlayed: number;
  timeTaken: number;
  avgAccuracy: number;
  attemptsTaken: number;
  gameStatus: string;
  createdAt?: string;
  completedAt?: string;
  rounds: AdminRound[];
};

function isRecord(value: unknown): value is Record<string, unknown> {
  return value != null && typeof value === "object" && !Array.isArray(value);
}

function toOptionalString(value: unknown): string | undefined {
  if (typeof value !== "string") return undefined;
  const trimmed = value.trim();
  return trimmed.length > 0 ? trimmed : undefined;
}

function toFiniteNumber(value: unknown, fallback = 0): number {
  if (typeof value === "number") {
    return Number.isFinite(value) ? value : fallback;
  }

  if (typeof value === "string") {
    const trimmed = value.trim();
    if (!trimmed) return fallback;
    const parsed = Number(trimmed);
    return Number.isFinite(parsed) ? parsed : fallback;
  }

  return fallback;
}

function toIsoString(value: unknown): string | undefined {
  if (value == null || value === "") return undefined;

  if (value instanceof Date) {
    return Number.isNaN(value.getTime()) ? undefined : value.toISOString();
  }

  if (typeof value === "number") {
    if (!Number.isFinite(value) || value <= 0) return undefined;
    const parsed = new Date(value);
    return Number.isNaN(parsed.getTime()) ? undefined : parsed.toISOString();
  }

  if (typeof value === "string") {
    const trimmed = value.trim();
    if (!trimmed) return undefined;
    const parsed = new Date(trimmed);
    return Number.isNaN(parsed.getTime()) ? undefined : parsed.toISOString();
  }

  return undefined;
}

function getFallbackPlayerId(
  email: string | undefined,
  name: string,
  createdAt: string | undefined,
  completedAt: string | undefined,
  index: number
) {
  const base = email ?? name.toLowerCase().replace(/\s+/g, "-");
  return `${base}-${createdAt ?? completedAt ?? `idx-${index}`}`;
}

function toPlayerId(value: unknown, fallback: string): string {
  if (typeof value === "string") {
    const trimmed = value.trim();
    if (trimmed) return trimmed;
  }

  if (value != null && typeof value === "object") {
    const maybeToString = (value as { toString?: unknown }).toString;
    if (typeof maybeToString === "function") {
      const result = maybeToString.call(value);
      if (typeof result === "string") {
        const trimmed = result.trim();
        if (trimmed && trimmed !== "[object Object]") return trimmed;
      }
    }
  }

  return fallback;
}

function normalizeRounds(value: unknown): AdminRound[] {
  if (!Array.isArray(value)) return [];

  return value.flatMap((round) => {
    if (!isRecord(round)) return [];

    return [
      {
        round: toFiniteNumber(round.round),
        attempts: toFiniteNumber(round.attempts),
        score: toFiniteNumber(round.score),
        prompt: round.prompt ?? null,
        output: toOptionalString(round.output) ?? "",
      },
    ];
  });
}

function isCompletedStatus(status: string) {
  return status === "COMPLETED" || status === "COMPLETED_WITH_BONUS";
}

function normalizeAdminPlayer(
  doc: RawAdminPlayerDoc,
  index = 0
): NormalizedAdminPlayer {
  const name = toOptionalString(doc.name) ?? "Unknown";
  const email = toOptionalString(doc.email);
  const createdAt = toIsoString(doc.createdAt);
  const completedAt = toIsoString(doc.completedAt);
  const fallbackId = getFallbackPlayerId(
    email,
    name,
    createdAt,
    completedAt,
    index
  );

  return {
    id: toPlayerId(doc._id, fallbackId),
    name,
    email,
    roundsPlayed: Math.max(0, toFiniteNumber(doc.roundsPlayed)),
    timeTaken: Math.max(0, toFiniteNumber(doc.timeTaken)),
    avgAccuracy: toFiniteNumber(doc.avgAccuracy),
    attemptsTaken: Math.max(0, toFiniteNumber(doc.attemptsTaken)),
    gameStatus: toOptionalString(doc.gameStatus) ?? "IN_PROGRESS",
    createdAt,
    completedAt,
    rounds: normalizeRounds(doc.rounds),
  };
}

export function toAdminPlayerSummary(
  doc: RawAdminPlayerDoc,
  index = 0
): AdminPlayerSummary {
  return toAdminPlayerSummaryShape(normalizeAdminPlayer(doc, index));
}

export function toAdminPlayerExport(
  doc: RawAdminPlayerDoc,
  index = 0
): AdminPlayerExport {
  return toAdminPlayerExportShape(normalizeAdminPlayer(doc, index));
}

function getPlayerAttemptCount(player: Player): number {
  if (typeof player.totalAttempts === "number") return player.totalAttempts;
  if (typeof player.attempts === "number") return player.attempts;

  return Object.values(player.attemptsPerRound ?? {}).reduce(
    (sum, value) => sum + toFiniteNumber(value),
    0
  );
}

function getPlayerGameStatus(player: Player): GameStatus | "IN_PROGRESS" {
  if (player.gameStatus) return player.gameStatus;
  return player.completed ? "COMPLETED" : "IN_PROGRESS";
}

function normalizeStoredPlayer(player: Player, index: number): NormalizedAdminPlayer {
  const name = toOptionalString(player.name) ?? "Unknown";
  const email = toOptionalString(player.email);
  const createdAt = toIsoString(player.startedAt);
  const completedAt = toIsoString(player.completedAt);

  return {
    id: toPlayerId(
      player.playerId,
      getFallbackPlayerId(email, name, createdAt, completedAt, index)
    ),
    name,
    email,
    roundsPlayed: Math.max(0, toFiniteNumber(player.roundsPlayed)),
    timeTaken:
      typeof player.completedAt === "number" && player.completedAt > 0
        ? Math.max(0, player.completedAt - player.startedAt)
        : 0,
    avgAccuracy: toFiniteNumber(player.averageScore),
    attemptsTaken: Math.max(0, getPlayerAttemptCount(player)),
    gameStatus: getPlayerGameStatus(player),
    createdAt,
    completedAt,
    rounds: [],
  };
}

function toAdminPlayerSummaryShape(player: NormalizedAdminPlayer): AdminPlayerSummary {
  return {
    playerId: player.id,
    name: player.name,
    email: player.email,
    roundsPlayed: player.roundsPlayed,
    timeTakenSec: Math.round(
      player.timeTaken > 10_000 ? player.timeTaken / 1000 : player.timeTaken
    ),
    averageScore: player.avgAccuracy,
    attemptsUsed: player.attemptsTaken,
    completed: isCompletedStatus(player.gameStatus),
    gameStatus: player.gameStatus,
  };
}

function toAdminPlayerExportShape(player: NormalizedAdminPlayer): AdminPlayerExport {
  return {
    _id: player.id,
    name: player.name,
    email: player.email,
    roundsPlayed: player.roundsPlayed,
    timeTaken: player.timeTaken,
    avgAccuracy: player.avgAccuracy,
    attemptsTaken: player.attemptsTaken,
    gameStatus: player.gameStatus,
    createdAt: player.createdAt,
    completedAt: player.completedAt,
    rounds: player.rounds,
  };
}

export function sortAdminPlayers<
  T extends {
    roundsPlayed: number;
    avgAccuracy: number;
    timeTaken: number;
    attemptsTaken: number;
    name: string;
  },
>(players: T[]): T[] {
  return players.sort((a, b) =>
    compareCompetitiveStanding(
      {
        roundsPlayed: a.roundsPlayed,
        averageScore: a.avgAccuracy,
        timeTakenMs: a.timeTaken,
        attempts: a.attemptsTaken,
        name: a.name,
      },
      {
        roundsPlayed: b.roundsPlayed,
        averageScore: b.avgAccuracy,
        timeTakenMs: b.timeTaken,
        attempts: b.attemptsTaken,
        name: b.name,
      }
    )
  );
}

function getMemoryPlayers(): Player[] {
  return rankPlayers(getPlayers().filter((player) => !isAdminEmail(player.email)));
}

export function getFallbackAdminPlayerSummaries(): AdminPlayerSummary[] {
  return getMemoryPlayers().map((player, index) =>
    toAdminPlayerSummaryShape(normalizeStoredPlayer(player, index))
  );
}

export function getFallbackAdminPlayerExports(): AdminPlayerExport[] {
  return getMemoryPlayers().map((player, index) =>
    toAdminPlayerExportShape(normalizeStoredPlayer(player, index))
  );
}

File: PromptWars_Aetherion_project/PromptWars_Aetherion/lib/admin_players.py
from __future__ import annotations
from datetime import datetime, timezone
from typing import Any, Optional
from PromptWars_Aetherion.lib.admin import is_admin_email
from PromptWars_Aetherion.lib.player_store import get_players
from PromptWars_Aetherion.lib.ranking import rank_players, compare_competitive_standing
from PromptWars_Aetherion.lib.types import Player, GameStatus


class AdminRound:
    def __init__(self, round: int, attempts: int, score: float, prompt: Any, output: str):
        self.round = round
        self.attempts = attempts
        self.score = score
        self.prompt = prompt
        self.output = output

    def to_dict(self) -> dict:
        return {"round": self.round, "attempts": self.attempts, "score": self.score, "prompt": self.prompt, "output": self.output}


class AdminPlayerSummary:
    def __init__(self, playerId: str, name: str, email: Optional[str], roundsPlayed: int,
                 timeTakenSec: int, averageScore: float, attemptsUsed: int, completed: bool, gameStatus: Optional[str]):
        self.playerId = playerId
        self.name = name
        self.email = email
        self.roundsPlayed = roundsPlayed
        self.timeTakenSec = timeTakenSec
        self.averageScore = averageScore
        self.attemptsUsed = attemptsUsed
        self.completed = completed
        self.gameStatus = gameStatus

    def to_dict(self) -> dict:
        return {k: v for k, v in self.__dict__.items() if v is not None}


class AdminPlayerExport:
    def __init__(self, _id: str, name: str, email: Optional[str], roundsPlayed: int,
                 timeTaken: int, avgAccuracy: float, attemptsTaken: int, gameStatus: str,
                 createdAt: Optional[str], completedAt: Optional[str], rounds: list[dict]):
        self._id = _id
        self.name = name
        self.email = email
        self.roundsPlayed = roundsPlayed
        self.timeTaken = timeTaken
        self.avgAccuracy = avgAccuracy
        self.attemptsTaken = attemptsTaken
        self.gameStatus = gameStatus
        self.createdAt = createdAt
        self.completedAt = completedAt
        self.rounds = rounds

    def to_dict(self) -> dict:
        d = {"_id": self._id, "name": self.name, "roundsPlayed": self.roundsPlayed,
             "timeTaken": self.timeTaken, "avgAccuracy": self.avgAccuracy,
             "attemptsTaken": self.attemptsTaken, "gameStatus": self.gameStatus,
             "rounds": self.rounds}
        if self.email:
            d["email"] = self.email
        if self.createdAt:
            d["createdAt"] = self.createdAt
        if self.completedAt:
            d["completedAt"] = self.completedAt
        return d


def _to_finite(value: Any, fallback: float = 0.0) -> float:
    try:
        v = float(value)
        return v if v == v and abs(v) != float("inf") else fallback
    except (TypeError, ValueError):
        return fallback


def _to_iso(value: Any) -> Optional[str]:
    if value is None or value == "":
        return None
    if isinstance(value, datetime):
        return value.isoformat() if not (value.year == 1 and value.month == 1) else None
    if isinstance(value, (int, float)):
        v = float(value)
        if v <= 0 or v != v:
            return None
        return datetime.fromtimestamp(v / 1000, tz=timezone.utc).isoformat()
    if isinstance(value, str):
        try:
            return datetime.fromisoformat(value.replace("Z", "+00:00")).isoformat()
        except ValueError:
            return None
    return None


def _to_str(value: Any) -> Optional[str]:
    if not isinstance(value, str):
        return None
    v = value.strip()
    return v if v else None


def _normalize_rounds(value: Any) -> list[dict]:
    if not isinstance(value, list):
        return []
    result = []
    for r in value:
        if not isinstance(r, dict):
            continue
        result.append({
            "round": int(_to_finite(r.get("round"))),
            "attempts": int(_to_finite(r.get("attempts"))),
            "score": _to_finite(r.get("score")),
            "prompt": r.get("prompt"),
            "output": _to_str(r.get("output")) or "",
        })
    return result


def _is_completed_status(status: str) -> bool:
    return status in ("COMPLETED", "COMPLETED_WITH_BONUS")


def _normalize_doc(doc: dict, index: int = 0) -> dict:
    name = _to_str(doc.get("name")) or "Unknown"
    email = _to_str(doc.get("email"))
    created_at = _to_iso(doc.get("createdAt"))
    completed_at = _to_iso(doc.get("completedAt"))
    raw_id = doc.get("_id")
    if isinstance(raw_id, str) and raw_id.strip():
        pid = raw_id.strip()
    elif raw_id is not None:
        pid = str(raw_id)
        if pid == "[object Object]":
            pid = f"{email or name.lower().replace(' ', '-')}-{created_at or completed_at or f'idx-{index}'}"
    else:
        pid = f"{email or name.lower().replace(' ', '-')}-{created_at or completed_at or f'idx-{index}'}"

    return {
        "id": pid,
        "name": name,
        "email": email,
        "roundsPlayed": max(0, int(_to_finite(doc.get("roundsPlayed")))),
        "timeTaken": max(0, int(_to_finite(doc.get("timeTaken")))),
        "avgAccuracy": _to_finite(doc.get("avgAccuracy")),
        "attemptsTaken": max(0, int(_to_finite(doc.get("attemptsTaken")))),
        "gameStatus": _to_str(doc.get("gameStatus")) or "IN_PROGRESS",
        "createdAt": created_at,
        "completedAt": completed_at,
        "rounds": _normalize_rounds(doc.get("rounds")),
    }


def to_admin_player_summary(doc: dict, index: int = 0) -> AdminPlayerSummary:
    n = _normalize_doc(doc, index)
    time_ms = n["timeTaken"]
    time_sec = round(time_ms / 1000 if time_ms > 10_000 else time_ms)
    return AdminPlayerSummary(
        playerId=n["id"],
        name=n["name"],
        email=n["email"],
        roundsPlayed=n["roundsPlayed"],
        timeTakenSec=time_sec,
        averageScore=n["avgAccuracy"],
        attemptsUsed=n["attemptsTaken"],
        completed=_is_completed_status(n["gameStatus"]),
        gameStatus=n["gameStatus"],
    )


def to_admin_player_export(doc: dict, index: int = 0) -> AdminPlayerExport:
    n = _normalize_doc(doc, index)
    return AdminPlayerExport(
        _id=n["id"],
        name=n["name"],
        email=n["email"],
        roundsPlayed=n["roundsPlayed"],
        timeTaken=n["timeTaken"],
        avgAccuracy=n["avgAccuracy"],
        attemptsTaken=n["attemptsTaken"],
        gameStatus=n["gameStatus"],
        createdAt=n["createdAt"],
        completedAt=n["completedAt"],
        rounds=n["rounds"],
    )


def sort_admin_players(players: list[dict]) -> list[dict]:
    def _sort_key(p: dict):
        score = _to_finite(p.get("avgAccuracy"))
        time_ms = int(_to_finite(p.get("timeTaken")))
        rounds = int(_to_finite(p.get("roundsPlayed")))
        attempts = int(_to_finite(p.get("attemptsTaken")))
        name = _to_str(p.get("name")) or ""

        from PromptWars_Aetherion.lib.ranking import get_accuracy_time_composite_score
        perf = get_accuracy_time_composite_score(score, time_ms)
        avg_att = attempts / rounds if rounds > 0 else attempts
        return (-rounds, -perf, avg_att, attempts, name)

    return sorted(players, key=_sort_key)


def _get_memory_players() -> list[Player]:
    return rank_players([p for p in get_players() if not is_admin_email(p.email)])


def get_fallback_admin_player_summaries() -> list[dict]:
    return [
        to_admin_player_summary({
            "_id": p.playerId,
            "name": p.name,
            "email": p.email,
            "roundsPlayed": p.roundsPlayed,
            "timeTaken": max(0, (p.completedAt or 0) - p.startedAt) if p.completedAt else 0,
            "avgAccuracy": p.averageScore,
            "attemptsTaken": sum(p.attemptsPerRound.values()) if p.attemptsPerRound else 0,
            "gameStatus": p.gameStatus or ("COMPLETED" if p.completed else "IN_PROGRESS"),
            "createdAt": p.startedAt,
            "completedAt": p.completedAt,
            "rounds": [],
        }, i).to_dict()
        for i, p in enumerate(_get_memory_players())
    ]


def get_fallback_admin_player_exports() -> list[dict]:
    return [
        to_admin_player_export({
            "_id": p.playerId,
            "name": p.name,
            "email": p.email,
            "roundsPlayed": p.roundsPlayed,
            "timeTaken": max(0, (p.completedAt or 0) - p.startedAt) if p.completedAt else 0,
            "avgAccuracy": p.averageScore,
            "attemptsTaken": sum(p.attemptsPerRound.values()) if p.attemptsPerRound else 0,
            "gameStatus": p.gameStatus or ("COMPLETED" if p.completed else "IN_PROGRESS"),
            "createdAt": p.startedAt,
            "completedAt": p.completedAt,
            "rounds": [],
        }, i).to_dict()
        for i, p in enumerate(_get_memory_players())
    ]
